File: realtube-python/app/routers/metrics.py
"""Prometheus metrics endpoint and collectors.

Design: infrastructure-design.md section 22.

Metrics exposed:
  realtube_votes_total{category}                     Counter
  realtube_api_request_duration_seconds{endpoint,method,status}  Histogram
  realtube_requests_in_flight                        Gauge
  realtube_cache_hits_total                          Counter
  realtube_cache_misses_total                        Counter
  realtube_score_recalculation_duration_seconds       Histogram
  realtube_db_connection_pool_active                  Gauge (callback)
  realtube_db_connection_pool_idle                    Gauge (callback)
"""

import time
import re

from fastapi import APIRouter, Request
from fastapi.responses import PlainTextResponse
from prometheus_client import (
    Counter,
    Gauge,
    Histogram,
    generate_latest,
    CONTENT_TYPE_LATEST,
)
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import Response

router = APIRouter(tags=["metrics"])

# ── Collectors ──

votes_total = Counter(
    "realtube_votes_total",
    "Total votes submitted, by category.",
    ["category"],
)

request_duration = Histogram(
    "realtube_api_request_duration_seconds",
    "HTTP request duration in seconds, by endpoint and method.",
    ["endpoint", "method", "status"],
)

requests_in_flight = Gauge(
    "realtube_requests_in_flight",
    "Number of HTTP requests currently being served.",
)

cache_hits = Counter(
    "realtube_cache_hits_total",
    "Total Redis cache hits.",
)

cache_misses = Counter(
    "realtube_cache_misses_total",
    "Total Redis cache misses.",
)

score_recalc_duration = Histogram(
    "realtube_score_recalculation_duration_seconds",
    "Duration of video score recalculations.",
)

db_pool_active = Gauge(
    "realtube_db_connection_pool_active",
    "Number of active database connections.",
)

db_pool_idle = Gauge(
    "realtube_db_connection_pool_idle",
    "Number of idle database connections.",
)

# Patterns for sanitizing dynamic path segments to avoid label cardinality explosion.
_ENDPOINT_PATTERNS = [
    (re.compile(r"^/api/videos/.+$"), "/api/videos/:hashPrefix"),
    (re.compile(r"^/api/channels/.+$"), "/api/channels/:channelId"),
    (re.compile(r"^/api/users/.+$"), "/api/users/:userId"),
]


def _sanitize_endpoint(path: str) -> str:
    for pattern, replacement in _ENDPOINT_PATTERNS:
        if pattern.match(path):
            return replacement
    return path


# ── Metrics endpoint ──


@router.get("/metrics")
async def metrics_endpoint(request: Request):
    """Serve Prometheus metrics."""
    # Update DB pool gauges from live pool stats
    pool = getattr(request.app.state, "db_pool", None)
    if pool is not None:
        db_pool_active.set(pool.get_size() - pool.get_idle_size())
        db_pool_idle.set(pool.get_idle_size())

    output = generate_latest()
    return PlainTextResponse(content=output, media_type=CONTENT_TYPE_LATEST)


# ── Middleware ──


class PrometheusMiddleware(BaseHTTPMiddleware):
    """Records request duration and in-flight count for Prometheus."""

    async def dispatch(self, request: Request, call_next) -> Response:
        # Don't instrument the /metrics endpoint itself
        if request.url.path == "/metrics":
            return await call_next(request)

        requests_in_flight.inc()
        start = time.perf_counter()

        response = await call_next(request)

        duration = time.perf_counter() - start
        endpoint = _sanitize_endpoint(request.url.path)
        request_duration.labels(
            endpoint=endpoint,
            method=request.method,
            status=str(response.status_code),
        ).observe(duration)
        requests_in_flight.dec()

        return response

File: realtube-go/internal/handler/metrics.go
package handler

import (
	"strconv"
	"time"

	"github.com/gofiber/fiber/v3"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/valyala/fasthttp/fasthttpadaptor"
)

// Metrics holds all Prometheus collectors for the RealTube Go backend.
var Metrics = struct {
	VotesTotal           *prometheus.CounterVec
	RequestDuration      *prometheus.HistogramVec
	DBPoolActive         prometheus.GaugeFunc
	DBPoolIdle           prometheus.GaugeFunc
	RequestsInFlight     prometheus.Gauge
	CacheHits            prometheus.Counter
	CacheMisses          prometheus.Counter
	ScoreRecalcDuration  prometheus.Histogram
}{}

// InitMetrics registers all Prometheus metrics. Call once at startup.
func InitMetrics(pool *pgxpool.Pool) {
	Metrics.VotesTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Name: "realtube_votes_total",
			Help: "Total votes submitted, by category.",
		},
		[]string{"category"},
	)

	Metrics.RequestDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "realtube_api_request_duration_seconds",
			Help:    "HTTP request duration in seconds, by endpoint and method.",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"endpoint", "method", "status"},
	)

	Metrics.RequestsInFlight = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "realtube_requests_in_flight",
			Help: "Number of HTTP requests currently being served.",
		},
	)

	Metrics.CacheHits = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "realtube_cache_hits_total",
			Help: "Total Redis cache hits.",
		},
	)

	Metrics.CacheMisses = prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "realtube_cache_misses_total",
			Help: "Total Redis cache misses.",
		},
	)

	Metrics.ScoreRecalcDuration = prometheus.NewHistogram(
		prometheus.HistogramOpts{
			Name:    "realtube_score_recalculation_duration_seconds",
			Help:    "Duration of video score recalculations.",
			Buckets: prometheus.DefBuckets,
		},
	)

	// DB pool gauges — read live stats from pgxpool
	if pool != nil {
		Metrics.DBPoolActive = prometheus.NewGaugeFunc(
			prometheus.GaugeOpts{
				Name: "realtube_db_connection_pool_active",
				Help: "Number of active database connections.",
			},
			func() float64 {
				return float64(pool.Stat().AcquiredConns())
			},
		)

		Metrics.DBPoolIdle = prometheus.NewGaugeFunc(
			prometheus.GaugeOpts{
				Name: "realtube_db_connection_pool_idle",
				Help: "Number of idle database connections.",
			},
			func() float64 {
				return float64(pool.Stat().IdleConns())
			},
		)

		prometheus.MustRegister(Metrics.DBPoolActive)
		prometheus.MustRegister(Metrics.DBPoolIdle)
	}

	prometheus.MustRegister(
		Metrics.VotesTotal,
		Metrics.RequestDuration,
		Metrics.RequestsInFlight,
		Metrics.CacheHits,
		Metrics.CacheMisses,
		Metrics.ScoreRecalcDuration,
	)
}

// MetricsMiddleware records request duration and in-flight count for Prometheus.
func MetricsMiddleware() fiber.Handler {
	return func(c fiber.Ctx) error {
		// Don't instrument the /metrics endpoint itself
		if c.Path() == "/metrics" {
			return c.Next()
		}

		// Copy path and method into owned strings BEFORE c.Next() — Fiber
		// returns slices backed by the fasthttp buffer which can be reused
		// or overwritten by handlers (especially fasthttpadaptor).
		path := string([]byte(c.Path()))
		method := string([]byte(c.Method()))
		endpoint := sanitizeEndpoint(path)

		Metrics.RequestsInFlight.Inc()
		start := time.Now()

		err := c.Next()

		duration := time.Since(start).Seconds()
		status := strconv.Itoa(c.Response().StatusCode())

		Metrics.RequestDuration.WithLabelValues(endpoint, method, status).Observe(duration)
		Metrics.RequestsInFlight.Dec()

		return err
	}
}

// sanitizeEndpoint normalizes paths to avoid cardinality explosion.
func sanitizeEndpoint(path string) string {
	switch {
	case len(path) > 12 && path[:12] == "/api/videos/":
		return "/api/videos/:hashPrefix"
	case len(path) > 14 && path[:14] == "/api/channels/":
		return "/api/channels/:channelId"
	case len(path) > 11 && path[:11] == "/api/users/":
		return "/api/users/:userId"
	default:
		return path
	}
}

// MetricsHandler serves the Prometheus /metrics endpoint via Fiber.
func MetricsHandler() fiber.Handler {
	httpHandler := fasthttpadaptor.NewFastHTTPHandler(promhttp.Handler())
	return func(c fiber.Ctx) error {
		httpHandler(c.RequestCtx())
		return nil
	}
}
